config/application.rb

```diff
@@ -208,5 +208,7 @@ class Application < Rails::Application
     config.generators.template_engine = :haml
 
     config.autoloader = :zeitwerk
+
+    config.action_view.form_with_generates_ids = true
```

Comment:
I think it's a good idea, and this is very handy when using capybara.
Using : 
```
fill_in "First Name", with: "Jane"
```

Reply:
:ok_hand:

Reply:
I don't understand the comment. Doesn't your example use the *label* "First Name" instead of an id? So what are we using the id for?

Maybe Rails is cleverer these days but the ids could be a problem when you have multiple forms on the page which can then generate the same ids. And that's invalid HTML.

Ah, now I see that the input is not nested within the label and therefore it needs an id to be referenced.

Another way around this is to include the input within the label. What do you think about this change?
```diff
--- a/app/views/split_checkout/_details.html.haml
+++ b/app/views/split_checkout/_details.html.haml
@@ -6,8 +6,9 @@
         = t("split_checkout.step1.your_details.title")
 
       %div.checkout-input
-        = bill_address.label :firstname, t("split_checkout.step1.your_details.first_name.label")
-        = bill_address.text_field :firstname, { placeholder: t("split_checkout.step1.your_details.first_name.placeholder") }
+        = bill_address.label :firstname do
+          = t("split_checkout.step1.your_details.first_name.label")
+          = bill_address.text_field :firstname, { placeholder: t("split_checkout.step1.your_details.first_name.placeholder") }
         = f.error_message_on "bill_address.firstname"
 
       %div.checkout-input
```

I'm not sure if we have the id problem somewhere or if your solution is much easier! :smile: I'm happy to go with your solution for now but I was wondering why it's not the Rails default and maybe there's a good reason for it.

Reply:
Without `config.action_view.form_with_generates_ids = true`, the generated html code is: 

```html
<div class="checkout-input">
  <label>First Name</label>
  <input placeholder="e.g. Jane" type="text" value="Jean-Baptiste" name="order[bill_address_attributes][firstname]">
</div>
```

When setting to true, html code has now `id` and `for` attribute (which is not only for testing purposes, but also for end user : by clicking in the label, the linked input file is focused): 
```html
<div class="checkout-input">
  <label for="order_bill_address_attributes_firstname">First Name</label>
  <input placeholder="e.g. Jane" type="text" value="Jean-Baptiste" name="order[bill_address_attributes][firstname]" id="order_bill_address_attributes_firstname">
</div>
```

(_NB_: the id seems to be smart generated as it contains all the path to the right attribute)

By putting the `<input />` inside the `<label />` tag, as you suggested it, we now have this html: 
```html
<label>
    First Name
    <input placeholder="e.g. Jane" type="text" value="Jean-Baptiste" name="order[bill_address_attributes][firstname]" class="focus-visible" data-focus-visible-added="">
</label>
```
which has the same behavior (from both capybara and end user point of view)

Since 5.2, it seems to be generated by default:
https://edgeguides.rubyonrails.org/5_2_release_notes.html#action-view-notable-changes
https://github.com/rails/rails/commit/260d6f112a0ffdbe03e6f5051504cb441c1e94cd

Happy to take solution `false` or `true` ...

Reply:
Yeah I wasn't sure about this at first glance, but as @jibees notes, this option is a recommended default for all Rails apps since Rails 5.2. If you generate a brand new Rails 6.1 app from scratch, this will be enabled.

Also, we don't have to worry too much about the effects of this elsewhere in the codebase, as it's only applied to the (relatively new) `#form_with` method (not `#form_for`). We've only just started using `#form_with` here in the split checkout and we don't currently use it elsewhere in the app :ok_hand:

Reply:
For a bit of background on the way successive Rails versions introduce new opt-in framework defaults which could break older apps: https://guides.rubyonrails.org/configuring.html#results-of-config-load-defaults